phac-nml/irida-next
# Edit metadata

## What does this PR do and why?
In this PR, the edit action has been added to the metadata listing ui. Users with access level >= Maintainer level will have access. The edit action will open a modal where users can edit the metadata key and/or value.

The metadata/edit_fields controller will initially receive the edit form param and validate it before calling the Samples::Metadata::UpdateService.

Note: The PR and changes take into account add and delete metadata which will be added afterwards. 
- `services/samples/metadata/fields` folder currently only contains `edit_service` but an `add_service` will be added.
- The table action ellipsis only contains edit (or nothing for analysis metadata), but delete will be added to the dropdown.

## Screenshots or screen recordings
New edit action:
![image](https://github.com/phac-nml/irida-next/assets/82407232/a4c6d055-0709-4a2e-b4fb-a0d191028b7a)

Edit modal:
![image](https://github.com/phac-nml/irida-next/assets/82407232/9d7fbbc7-bbfd-487c-90a0-7fd03bce3c57)

Successful update:
![image](https://github.com/phac-nml/irida-next/assets/82407232/a634c11e-dcaf-4071-a81c-52dac23b6c8d)

Unchanged metadata error:
![image](https://github.com/phac-nml/irida-next/assets/82407232/9e34bed4-c8b6-49fd-a955-56f68f0b9802)

Key already exists error:
![image](https://github.com/phac-nml/irida-next/assets/82407232/f0187aa6-59aa-4773-89ff-bb9fee70528a)


## How to set up and validate locally
1. Set up a sample with metadata, preferably with fields from both a user and analysis source.
2. Navigate to the sample metadata page
3. Edit both keys and values to ensure they update accordingly.
4. Ensure two keys with the same name cannot exist (including different cases/capitilization).
5. Ensure any metadata added by an analysis does not have the edit action available in the action dropdown.
6. Ensure users with access < Maintainer do not see the action column or action dropdown ellipsis.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
